crates/ide-assists/src/handlers/bool_to_enum.rs

```diff
@@ -0,0 +1,724 @@
+        if !def.ty(ctx.db()).is_bool() {
+            cov_mark::hit!(not_applicable_non_bool_const);
+            return None;
+        }
+
+        Some(BoolNodeData {
+            target_node: const_.syntax().clone(),
+            name: const_.name()?,
+            ty_annotation: const_.ty(),
+            initializer: const_.body(),
+            definition: Definition::Const(def),
+        })
+    } else if let Some(static_) = ctx.find_node_at_offset::<ast::Static>() {
+        let def = ctx.sema.to_def(&static_)?;
+        if !def.ty(ctx.db()).is_bool() {
+            cov_mark::hit!(not_applicable_non_bool_static);
+            return None;
+        }
+
+        Some(BoolNodeData {
+            target_node: static_.syntax().clone(),
+            name: static_.name()?,
+            ty_annotation: static_.ty(),
+            initializer: static_.body(),
+            definition: Definition::Static(def),
+        })
+    } else if let Some(field_name) = ctx.find_node_at_offset::<ast::Name>() {
+        let field = field_name.syntax().ancestors().find_map(ast::RecordField::cast)?;
+        if field.name()? != field_name {
+            return None;
+        }
+
+        let strukt = field.syntax().ancestors().find_map(ast::Struct::cast)?;
+        let def = ctx.sema.to_def(&field)?;
+        if !def.ty(ctx.db()).is_bool() {
+            cov_mark::hit!(not_applicable_non_bool_field);
+            return None;
+        }
+        Some(BoolNodeData {
+            target_node: strukt.syntax().clone(),
+            name: field_name,
+            ty_annotation: field.ty(),
+            initializer: None,
+            definition: Definition::Field(def),
+        })
+    } else {
+        None
+    }
+}
+
+fn replace_bool_expr(edit: &mut SourceChangeBuilder, expr: ast::Expr) {
+    let expr_range = expr.syntax().text_range();
+    let enum_expr = bool_expr_to_enum_expr(expr);
+    edit.replace(expr_range, enum_expr.syntax().text())
+}
+
+/// Converts an expression of type `bool` to one of the new enum type.
+fn bool_expr_to_enum_expr(expr: ast::Expr) -> ast::Expr {
+    let true_expr = make::expr_path(make::path_from_text("Bool::True")).clone_for_update();
+    let false_expr = make::expr_path(make::path_from_text("Bool::False")).clone_for_update();
+
+    if let ast::Expr::Literal(literal) = &expr {
+        match literal.kind() {
+            ast::LiteralKind::Bool(true) => true_expr,
+            ast::LiteralKind::Bool(false) => false_expr,
+            _ => expr,
+        }
+    } else {
+        make::expr_if(
+            expr,
+            make::tail_only_block_expr(true_expr),
+            Some(ast::ElseBranch::Block(make::tail_only_block_expr(false_expr))),
+        )
+        .clone_for_update()
+    }
+}
+
+/// Replaces all usages of the target identifier, both when read and written to.
+fn replace_usages(edit: &mut SourceChangeBuilder, usages: &UsageSearchResult) {
+    for (_, references) in usages.iter() {
```

Comment:
I think it would be nice to handle cross file and module references too, especially in the case of constants. I had a version before where I would just check to see if the `ctx` file is different than the usage file, then I would insert an import which seemed to work out alright for basic files but this doesn't handle modules within a file so I was wondering if there is another approach that would be nicer?

Reply:
Maybe handling a test like this could be nice?

```rust
    #[test]
    fn const_in_module() {
        check_assist(
            bool_to_enum,
            r#"
mod foo {
    pub const $0FOO: bool = true;
}

fn main() {
    if foo::FOO {
        println!("foo");
    }
}
"#,
            r#"
mod foo {
    #[derive(PartialEq, Eq)]
    pub enum Bool {
        True,
        False,
    }

    pub const FOO: Bool = Bool::True;
}

fn main() {
    use foo::Bool;

    if foo::FOO == Bool::True {
        println!("foo");
    }
}
"#,
        )
    }
```

Reply:
Two approaches may be:

1. Use `ImportScope::find_insert_use_container` to check if the definition and usage are in different containers and then do the appropriate `insert_use`.
2. Use `is_` methods for each of the variants, as done in `generate_enum_is_method` so that there are no imports needed.

Reply:
Handling it cross modules properly would be great yes, you might be able to check on `crates\ide-assists\src\handlers\extract_struct_from_enum_variant.rs` how things are done there. Iirc it needs to import stuff as well when rewriting other  files

Reply:
I see thanks for the pointer! I think I've got it to work in a similar way now using the import approach. The only problem is I can't seem to get the tabstop snippet to show up in the right place because of the added import, so I've just removed the tabstop for now.